lib/ansible/modules/packaging/os/pkgutil.py

```diff
@@ -64,146 +77,219 @@
+        # Fail with an explicit error when trying to "install" '*'
+        else:
+            module.fail_json(msg="Can not use state: present with name: *")
 
+    elif state in ['latest']:
+        if name != ['*']:
+            # Build list of packages that are either outdated or not installed
+            pkgs = packages_not_installed(module, name)
+            pkgs += packages_not_latest(module, name, site, update_catalog)
+            if pkgs != []:
+                (rc, out, err) = package_upgrade(module, pkgs, site, update_catalog, force)
+                if rc != 0:
+                    if err:
+                        msg = err
+                    else:
+                        msg = out
+                    module.fail_json(msg=msg)
+            # If the package list is empty that means all packages are installed and up to date
+            else:
+                module.exit_json(changed=False)
+        # When using latest for *
         else:
-            if not package_latest(module, name, site):
-                if module.check_mode:
-                    module.exit_json(changed=True)
-                (rc, out, err) = package_upgrade(module, name, site, update_catalog)
-                if len(out) > 75:
-                    out = out[:75] + '...'
+            # Check for packages that are actually outdated
+            pkgs = packages_not_latest(module, name, site, update_catalog)
+            if pkgs != []:
+                # If there are packages to update, just empty the list and run the command without it
+                # pkgutil logic is to update all when run without packages names
+                pkgs == []
+                (rc, out, err) = package_upgrade(module, pkgs, site, update_catalog, force)
                 if rc != 0:
                     if err:
                         msg = err
                     else:
                         msg = out
                     module.fail_json(msg=msg)
+            # If the package list comes up empty, everything is already up to date
+            else:
+                module.exit_json(changed=False)
 
-    elif state == 'absent':
-        if package_installed(module, name):
-            if module.check_mode:
-                module.exit_json(changed=True)
-            (rc, out, err) = package_uninstall(module, name)
-            if len(out) > 75:
-                out = out[:75] + '...'
+    elif state in ['absent', 'removed']:
+        # Build list of packages requested for removal that are actually present
+        pkgs = packages_installed(module, name)
+        if pkgs != []:
+            (rc, out, err) = package_uninstall(module, pkgs)
             if rc != 0:
                 if err:
                     msg = err
                 else:
                     msg = out
                 module.fail_json(msg=msg)
+        # If the list is empty, no packages need to be removed
+        else:
+            module.exit_json(changed=False)
```

Comment:
No `RETURN` is defined. I think you should add it and then add more parameters to the `module.exit_json()`. At least `state` should be passed to all `exit_json` calls.